File: py_solutions_21-30/Euler_21.py
# Calculate the sum of all amicable pairs below 10000 

from __future__ import print_function

import timeit
import functools

try:
    range = xrange
except NameError:
    pass

start = timeit.default_timer()

def mod_zero(dividend, divisor):
    return True if dividend % divisor == 0 else False

def euler_21():
    total = 0
    for i in range(1, 10000):
        p1 = sum(filter(functools.partial(mod_zero, i), range(1, i // 2 + 1)))  
        p2 = sum(filter(functools.partial(mod_zero, p1), range(1, p1 // 2 + 1)))
        if p2 == i != p1: 
            total += i
    return total


print("Answer: {}".format(euler_21()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))

File: C/c_solutions_21-30/Euler_21.c
// Calculate the sum of all the amicable pairs below 10000
#include "euler_util.h"

#define LIMIT 10000


int main(int argc, char *argv[])
{
    float start = timeit();

    int answer = 0; 

    for (int i=1, p1=0, p2=0; i < LIMIT; i++, p1=0, p2=0) {
        for (int j=1; j < (i / 2) + 1; j++) 
            p1 += i % j == 0 ? j : 0;
        for (int j=1; j < (p1 / 2) + 1; j++)
            p2 += p1 % j == 0 ? j : 0;
        answer += (p2 == i) && (i != p1) ? i : 0;
    }

    float stop = timeit();

    printf("Answer: %d\n", answer);
    printf("Time: %f\n", stop - start);

    return 0;
}
